### DISCUSSION TITLE
Apostrophe 3.48.0 Rich-text-widget improvements, new enterprise modules, and fixes

### DISCUSSION THREAD

**Discussion by BoDonkey** [apostrophecms/apostrophe#4175] [Release Notes]
Hola Apostrophe Universe!

The Apostrophe team is getting back into the swing of things after our in-person meet-up at the Philly offices. We brainstormed and hacked on new potential uses for AI in Apostrophe including automatic translation, page creation, and cloud tag creation. Exciting times for content creation and software development!

In this release, we have made some improvements to the recently added rich-text-widget floating menu. Now it is easier to add text before and after tables and inline images, no matter when they are added to the page. Inline images added in the rich-text-widget now include the alt text from the media library.

We want to give a big shout-out to [NuntukDev](https://github.com/nuktukdev), a community contributor that closed a long-standing issue in Apostrophe core! 🎉 We are always interested in working with the community to make our product better, either through contributed changes to the code, ideas for enhancement on our [product roadmap](https://portal.productboard.com/apostrophecms/1-product-roadmap/tabs/2-planned), or other feedback in our [Discord](https://discord.com/invite/XkbRNq7).

As part of this release we have also added a new Pro module – the [`@apostrophecms-pro/signup`](https://apostrophecms.com/extensions/account-signup) module allows users to easily sign-up for an account on your site. [Contact us](https://apostrophecms.com/contact-us) or visit our new [pricing page](https://apostrophecms.com/pricing) to learn more.

Moving into the next cycle, our team is working on making it easier to start new projects with a unique look and add pages to existing projects. We have been working with partner agencies to design templates in a variety of niches, like e-commerce, to help bootstrap your next project more quickly!

### Adds

* For performance, add `apos.modules['piece-type']getManagerApiProjection` method to reduce the amount of data returned in the manager
    modal. The projection will contain the fields returned in the method in addition to the existing manager modal
    columns.
* Add `apos.schema.getRelationshipQueryBuilderChoicesProjection` method to set the projection used in
    `apos.schema.relationshipQueryBuilderChoices`.
* Rich-text inline images now copies the `alt` attribute from the original image from the Media Library.

### Changes

* You can set `max-width` on `.apos-rich-text-toolbar__inner` to define the width of the rich-text toolbar. It will now
    flow on multiple lines if needed.
* The `utilityRail` prop of `AposSchema` now defaults to `false`, removing
the need to explicitly pass it in almost all contexts.
* Mark `apos.modules['doc-type']` methods `getAutocompleteTitle`, `getAutocompleteProjection` and `autocomplete` as
    deprecated. Our admin UI does not use them, it uses the `autocomplete('...')` query builder.
    More info at https://v3.docs.apostrophecms.org/reference/query-builders.html#autocomplete'.
* Print a warning with a clear explanation if a module's `index.js` file contains
no `module.exports` object (often due to a typo), or it is empty.

### Fixes

* Now errors and exits when a piece-type or widget-type module has a field object with the property `type`. Thanks to [NuktukDev](https://github.com/nuktukdev) for this contribution.
* Add a default page type value to prevent the dropdown from containing an empty value.
  
### Fixes

* Does not display `(undefined)` if the outer array item is new when an array is nested in another array and the parent object appears in the breadcrumb trail while editing the nested array.

### [`@apostrophecms/ai-helper`](https://github.com/apostrophecms/ai-helper) 1.0.0-beta.2

This module enhances Apostrophe with AI-driven helpers. Currently this module offers:

- A button to generate an image from a text prompt via OpenAI.
- An insert menu option to generate rich text from a prompt via OpenAI. 

See the [README](https://github.com/apostrophecms/ai-helper/blob/main/README.md) for more info.

### Changes

- Convert imported PNGs to JPEGs as this is more appropriate to web delivery of AI-generated, often realistic images. Reduces image download time to the browser by at least 80%.

### [`@apostrophecms-pro/palette`](https://apostrophecms.com/extensions/palette-3) 3.1.5

This module provides an in-context interface for changing the values of developer-set CSS properties. The values are stored in an Apostrophe piece singleton (like @apostrophecms/global) and applied to the site whenever the stylesheet link is included in a template. Adjusting values via the palette interface renders changes to the site instantly. [Contact us](https://apostrophecms.com/contact-us) or visit our new [pricing page](https://apostrophecms.com/pricing) to learn more.

### Fixes

* Fix error in browser console when handling `content-changed` event, check that `e.doc` exists.

### [`@apostrophecms-pro/advanced-permission`](https://apostrophecms.com/extensions/advanced-permission) 1.6.0

This module provides more granular control over content permission. It allows the creation of custom groups with proscribed abilities and the ability to assign users to as few or as many groups as desired. Users and groups can also be assigned per-document and per-locale permissions allowing for fine control of who can edit and publish any document. [Contact us](https://apostrophecms.com/contact-us) or visit our new [pricing page](https://apostrophecms.com/pricing) to learn more.

### Adds

* Add support for new apostrophe `apos.modules['piece-type'].getManagerApiProjection` method.
* Accurate license information.

### [`@apostrophecms-pro/signup`](https://apostrophecms.com/extensions/account-signup) 1.0.0-beta

A seamless solution that empowers users to easily create accounts on your website. Users can sign up in a few simple steps, complete with email verification and pre-determined user role and access level. [Contact us](https://apostrophecms.com/contact-us) or visit our new [pricing page](https://apostrophecms.com/pricing) to learn more.

* Initial release.

### [`apostrophe-forms`](https://www.npmjs.com/package/apostrophe-forms) 1.12.0

This module provides an easy solution to add forms to your Apostrophe 2 project, allowing forms to be saved to the database and easy email confirmation upon submission.

- Adds `Required` label to required fields since the `Optional` label is already in place.
- Fixes small class name typo from `apos-form-...` to `apos-forms-...`

### [`absolution`](https://www.npmjs.com/package/absolution) 1.0.4

`absolution` accepts HTML and a base URL, and returns HTML with absolute URLs. Great for generating valid RSS feeds.

### Adds

* Add test and documentation about adding custom self-closing tags.

**Comment by olukinni-029** [comment]
Awesome👏
